### DISCUSSION TITLE
How to have .json to bit filed factories?

### DISCUSSION THREAD

**Discussion by JiveyGuy** [jam1garner/binrw#216] [Q&A] [answered]
If I have a .json that describes a bitfield format as the following do:

```json
// data file part 1
{
    "enable_bit_1"  : "bool",
    "enable_bit_2"  : "bool",
    "data_1"        : "B4",
    "enum_2"        : "enum {opt_1, opt_2, opt_3, opt_4}",
    "data_2"        : "Signed B5"
}

// data file part 2
{
    "entire_word" : "B32"
}
```

How can that be done since we define everything with structs at compile time? Would we need a different `impl fn` for each format?

**Comment by kitlith** [comment, marked as answer]
First off, binrw doesn't deal with bitfields -- you'll want another crate for that. (For example, modular bitfield, deku, nom, etc.)

Second off, how you implement things depends on what your goal is.

If you know what formats you're going to be dealing with at compile time, and want to use the json as your source of truth, you could write a proc-macro (or other tool) that generates structs declarations that use <crate of your choice> based off of your json file, and then your code could use those generated types to parse files.

On the other hand, if it's supposed to be a runtime definition of a data format... essentially what you'd do is parse each individual data type out of the bytestream. so instead of parsing a "DataFile" struct, you'd parse a bool, then another bool, then a 4 bit integer, etc, and which type you parse out would be controlled by the json file. In this instance, any custom types you'd define would be types present in your spec -- for example, a custom variable length integer. I'll reiterate here that since you're working with bit fields, you should consider a crate that can handle bit streams, such as nom or deku. modular-bitfield is probably a poor match for the runtime usecase.

**Reply by JiveyGuy** [reply]
I see thanks, this gives a good general idea of how to do this. I think the generation of structs at compile time sounds like the best approach. I will go check out modular bitfields.

**Reply by kitlith** [reply]
To add to that list of crates you might be interested in: bilge and bitbybit. I had forgotten to list them before, but I had to deal with bitfields again myself in the last few days.

**Reply by JiveyGuy** [reply]
Will be sure to look at those, thanks again! I might revisit this topic sometime as a side project but I will be using modular bitfields for a work task that is pretty urgent.
